### PR TITLE
create arbiter bricks sets in reverse order

### PR DESCRIPTION
**Pull request by phlogistonjohn**
### What does this PR achieve? Why do we need it?

Creating the arbiter bricks in reverse order (starting with the 3rd
brick in the brick set and working back to the 1st) means we create the arbiter brick first
and this prevents the data bricks from "blocking" the arbiter brick
being placed on a particular node. This makes placing the
arbiter bricks more reliable.

### Does this PR fix issues?

See [rhbz #1572561](https://bugzilla.redhat.com/show_bug.cgi?id=1572561)

### Notes for the reviewer

This PR relies on the changes in PR #1171 so that tests can pass. Please review that PR first.

### CODE REVIEW

File: apps/glusterfs/placer_arbiter.go [hunk 1]
```diff
@@ -103,6 +103,12 @@ func (bp *ArbiterBrickPlacer) PlaceAll(
 		if err != nil {
 			return r, err
 		}
+		if bs.IsSparse() {
```
Comment (nixpanic) [inline apps/glusterfs/placer_arbiter.go]: The 'Sparse' naming might be confusing me. Why would a SparseBrick/Device be an error here?

File: apps/glusterfs/brick_allocate.go [hunk 2]
```diff
@@ -28,6 +28,13 @@ func NewBrickSet(s int) *BrickSet {
 	return &BrickSet{SetSize: s, Bricks: []*BrickEntry{}}
 }
 
+func NewSparseBrickSet(s int) *BrickSet {
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply apps/glusterfs/brick_allocate.go]: Heh. I always get in trouble with names. It's not that the brick set is sparse at the gluster level, its that the brick set object in heketi starts off "sparse" in that the array holds nulls but is sized up to the expected brick set set, rather than being an empty array.

It's not arbiter specific because I had wanted something like this earlier for some other re-factoring I did, but I convinced myself not to do it then. I will try to see if I can come up with a better name quickly.

File: apps/glusterfs/placer_arbiter.go [hunk 3]
```diff
@@ -171,15 +177,20 @@ func (bp *ArbiterBrickPlacer) newSets(
...
 	}
 	defer dscan.Close()
 
-	for index := 0; index < ssize; index++ {
+	// work backwards from the last item in the brick set (typically index 2)
+	// in order to place the most special brick, the arbiter brick, first.
+	// Placing the arbiter brick first means we get a more reliable distribution
+	// of bricks because the nodes will not be "reserved" by the two data
+	// bricks before we try to place the arbiter brick.
+	for index := ssize - 1; index >= 0; index-- {
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply apps/glusterfs/placer_arbiter.go]: It just means taken. I should have just wrote that. I will update the comment.

File: apps/glusterfs/brick_allocate.go [hunk 4]
```diff
@@ -24,10 +24,23 @@ type BrickSet struct {
...
+// NewBrickSet creates an empty brick set tracking object with
+// the expected final size of s.
 func NewBrickSet(s int) *BrickSet {
 	return &BrickSet{SetSize: s, Bricks: []*BrickEntry{}}
 }
 
+// NewSparseBrickSet creates a brick set tracking object with
+// the expected final size of s and the brick tracking slice
+// pre-sized to s. This means that the array can be directly
+// indexed but may contains null brick entries.
```
Comment (ansiwen) [inline apps/glusterfs/brick_allocate.go]: typo: ...may contain null...

File: apps/glusterfs/brick_allocate.go [hunk 5]
```diff
@@ -24,10 +24,23 @@ type BrickSet struct {
...
+func NewSparseBrickSet(s int) *BrickSet {
+	return &BrickSet{
+		SetSize: s,
+		Bricks:  make([]*BrickEntry, s),
```
Comment (ansiwen) [inline apps/glusterfs/brick_allocate.go]: I'm new to golang, but wouldn't it be better style, to make a new type from []*BrickEntry, like `type BrickSlice = []*BrickEntry` and use that instead?

File: apps/glusterfs/brick_allocate.go [hunk 6]
```diff
@@ -76,10 +112,23 @@ type DeviceSet struct {
...
 }
 
+// NewSparseDeviceSet creates a device set tracking object with
+// the expected final size of s and the device tracking slice
+// pre-sized to s. This means that the array can be directly
+// indexed but may contains null brick entries.
+func NewSparseDeviceSet(s int) *DeviceSet {
+	return &DeviceSet{
+		SetSize: s,
+		Devices: make([]*DeviceEntry, s),
```
Comment (ansiwen) [inline apps/glusterfs/brick_allocate.go]: Same as for []*BrickEntry

File: apps/glusterfs/placer_arbiter.go [hunk 7]
```diff
@@ -148,6 +160,11 @@ func (bp *ArbiterBrickPlacer) Replace(
 		wds.Insert(i, d)
 	}
 	aopts := newArbiterOpts(opts)
+	// this is a mildly hacky way to deal with the higher level replace
```
Comment (ansiwen) [inline apps/glusterfs/placer_arbiter.go]: I assume a non-hacky way is not feasible right now? ;-)

File: apps/glusterfs/placer_arbiter.go [hunk 8]
```diff
@@ -234,7 +256,7 @@ func (bp *ArbiterBrickPlacer) tryPlaceBrickOnDevice(
 
 	logger.Debug("Trying to place brick on device %v", device.Info.Id)
 
-	for i, b := range bs.Bricks {
+	for i, b := range bs.Contents() {
```
Comment (ansiwen) [inline apps/glusterfs/placer_arbiter.go]: Contents() is not a newly introduced function. Why the change?

File: apps/glusterfs/placer_arbiter.go (hunk 7 above)
Comment (phlogistonjohn) [reply apps/glusterfs/placer_arbiter.go]: Yup

File: apps/glusterfs/placer_arbiter.go (hunk 8 above)
Comment (phlogistonjohn) [reply apps/glusterfs/placer_arbiter.go]: Contents was introduced in the previous change. It returns a copy of the Bricks with all nulls removed for easy iteration.

File: apps/glusterfs/brick_allocate.go (hunk 5 above)
Comment (phlogistonjohn) [reply apps/glusterfs/brick_allocate.go]: It is certainly do-able... I just don't see the value of adding that now ATM. If we were to do it I'd add it to the file where BrickEntry was defined.

File: apps/glusterfs/placer_arbiter.go (hunk 1 above)
Comment (obnoxxx) [reply apps/glusterfs/placer_arbiter.go]: Because newSets() is supposed to return a fully populated set (i.e. one without nil entries).
These checks might be more natural at the end of newSets().
